### DISCUSSION TITLE
[Suggestion] Better performance with Erlang multisend?

### DISCUSSION THREAD

**Discussion by benbro** [rabbitmq/rabbitmq-server#12915] [Ideas] [answered]
### RabbitMQ series

4.0.x

### Operating system (distribution) used

Debian

### How is RabbitMQ deployed?

Community Docker image

### What would you like to suggest for a future version of RabbitMQ?

There is a [discussion](https://erlangforums.com/t/send-optimization-for-sending-to-many-processes/2200/16) and [PR](https://github.com/erlang/otp/pull/8072) for multisend support in Erlang - being able to send the same message to many pids. Might help for broadcasting the same message to many clients. Do you think it is relevant?

**Comment by ansd** [comment]
RabbitMQ uses the [delegate](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit_common/src/delegate.erl) module as custom solution to for example route a message to many queues. (The queues are hosted on a small number of different nodes. RabbitMQ clusters using Mnesia consist typically between 3 and 7 nodes). This custom solution has worked very well.
That said, RabbitMQ could make use of that PR in future if it gets merged into Erlang/OTP.

**Comment by michaelklishin** [comment, marked as answer]
@benbro thank you for bringing this to our attention.

Tanzu RabbitMQ had/has a prototype where it uses multiple TCP connections between nodes. It is very difficult to get right and the effects were not overwhelming, so the idea was dropped and hasn't resurfaced since.

As David said, RabbitMQ has its own solution for passing around messages that must be routed to N queues on a remote replica. It has been around since 2011 or so. You can tune the number of these delegates to match the number of cores. It does its job well and it is very easy to reason about (compared to the prototype above).

I did not read the entire discussion and parts of it focus on scheduler yielding, that is, node-local inter-process multi-cast. That's cheap compared to what `delegate` solves because going over the network is orders of magnitude more expensive than communicating with local processes. So far it looks like this would be a way to slightly optimize `delegate` when we can require the version of Erlang that will ship with this "multi-send" support, most likely Erlang 28 but could be 29. That's easily a good year or two from now.

For large fan-out scenarios you will likely be better off using a stream even then.
